src/embed/sync.rs

```diff
@@ -0,0 +1,128 @@
+use super::http as embed_http;
+use super::types::*;
+use nanoserde::{DeJson, DeJsonErr, SerJson};
+use std::fmt::Debug;
+
+#[cfg(default)]
+const USE_BROWSER_FETCH: bool = false;
+
+#[cfg(not(default))]
+const USE_BROWSER_FETCH: bool = true;
+
+pub async fn begin_sync(
+    begin_sync_req: &BeginSyncRequest,
+) -> Result<BeginSyncResponse, BeginSyncError> {
+    let _pull_resp = pull(begin_sync_req).await?;
+    // TODO do something with the response
+    Ok(BeginSyncResponse {})
+}
+
+#[derive(Debug)]
+pub enum BeginSyncError {
+    PullFailed(PullError),
+}
+
+impl From<PullError> for BeginSyncError {
```

Comment:
I think it is better overall to not use these From impls (despite the recommendations online) because they are one-to-one with source error. Whereas in real code you'll frequently have multiple places in the same function that cause the same error type, and will want to give them different enumerated tags. So I think we should just do `.map_err` at the call site. It also saves all this impl goop.

Reply:
hmmm. in this case i would find the proliferation of error types and map_err everywhere to be goop that gets in the way of understanding what is happening. i like that the errors here are 1-1 -- most of the source errors have the same type so with one from i have handled almost all of them. and i can still tell as a developer what went wrong because i'm capturing their debug output, but i'm not forced to enumerate all the separate ways this thing can fail. 

i have a hard time reconciling the feeling that the enumerated types are good at the sync level but not good here. i realize it might be a contradiction but it still feels different. maybe because in some sense this is the lowest (rust) level? like this is the bottommost place that a coherent thing is happening (an http fetch); at any lower level the things that are happening re in service of this level, so it feels natural to bottom out here for some (admittedly maybe contradictory) reason.

hmmmm.

Reply:
It seems pretty likely that somewhere in this sync code you'll have a function where http::Error happens twice in the same function no? It was exactly that kind of issue that led me away from this pattern.

Reply:
> not forced to enumerate all the separate ways this thing can fail.

This part is annoying with my proposal, I agree. I was hoping I could put something together like:

```
fetch(pull_url).map_err(FetchPullError)?
```

... and the FetchPullError would automatically go into an enum for the function. Then you wouldn't have to enumerate them, it would happen automatically.